### PR TITLE
Rek upload permissions

### PR DESCRIPTION
**Pull request by kilbergr**
## Description

A user should only be able to upload files to their own moves. This PR prevents any user other than the user who owns the move from being able to upload files.

## Reviewer Notes

I'm using a function from models/move `GetMovesForUserID`, which I know @macrael isn't totally comfortable with as a way to authenticate user ownership. We're going to discuss whether there's a better way, but just stay alert.

## Verification Steps

* [x] All tests pass.
* [x] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/blob/master/docs/backend.md#logging)
* [x] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/blob/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/156282848) for this change

### CODE REVIEW

File: pkg/handlers/uploads.go [hunk 1]
```diff
@@ -24,6 +24,21 @@ func payloadForUploadModel(upload models.Upload, url string) internalmessages.Up
 	}
 }
 
+// User uploading the file should be the same as the user who 'owns' a move
+func (h CreateUploadHandler) userOwnsMove(currentMoveID uuid.UUID, currentUserID uuid.UUID) bool {
+	moves, err := models.GetMovesForUserID(h.db, currentUserID)
```
Comment (ntwyman) [inline pkg/handlers/uploads.go]: Why doesn't this just load the specific move and then see if the user is the owner.?

File: pkg/handlers/uploads.go [hunk 2]
```diff
@@ -42,6 +57,10 @@ func (h CreateUploadHandler) Handle(params uploadop.CreateUploadParams) middlewa
 		h.logger.Panic("Invalid MoveID, this should never happen.")
```
Comment (ntwyman) [inline pkg/handlers/uploads.go]: This needs to change - should be 404

Comment (kilbergr) [reply pkg/handlers/uploads.go]: This is not a check that the document exists--it's a check that the uuid is properly formatted. I'm adding a 404 below.

File: pkg/handlers/uploads.go (hunk 1 above)
Comment (kilbergr) [reply pkg/handlers/uploads.go]: Changing the function used but yes.

File: pkg/models/document.go [hunk 3]
```diff
@@ -44,3 +44,25 @@ func (d *Document) Validate(tx *pop.Connection) (*validate.Errors, error) {
...
+// ValidateDocumentOwnership validates that a user owns the move that contains a document and that move and document both exist
+func ValidateDocumentOwnership(db *pop.Connection, userID uuid.UUID, moveID uuid.UUID, documentID uuid.UUID) (bool, bool, bool) {
+	docExists := false
+	moveExists := false
+	userOwns := false
+	var move Move
+	var document Document
+	err := db.Find(&document, documentID)
+	if err == nil {
+		docExists = true
```
Comment (macrael) [inline pkg/models/document.go]: You need to check that the document's moveID is equal to the moveID

File: pkg/models/document.go [hunk 4]
```diff
@@ -44,3 +44,25 @@ func (d *Document) Validate(tx *pop.Connection) (*validate.Errors, error) {
...
+	err = db.Find(&move, moveID)
+	if err == nil {
+		moveExists = true
+		// TODO: Handle case where more than one user is authorized to modify move
+		if uuid.Equal(move.UserID, userID) && uuid.Equal(document.MoveID, moveID) {
+			userOwns = true
+		}
+	}
+	return docExists, moveExists, userOwns
+}
```
Comment (macrael) [inline pkg/models/document.go]: Just to standardize, you might still just return a pair of variables here. One for the chain existing correctly (doc exists, belongs to move, move exists, belongs to user) and one for permissions (move does not belong to this user)

File: pkg/models/document.go (hunk 3 above)
Comment (kilbergr) [reply pkg/models/document.go]: See below on line 63. That is where I'm checking that ownership is as it should be. This is just to check whether each thing obj exists.

File: pkg/models/document.go (hunk 4 above)
Comment (kilbergr) [reply pkg/models/document.go]: OK.

File: pkg/handlers/uploads.go [hunk 5]
```diff
@@ -34,26 +34,37 @@ func (h CreateUploadHandler) Handle(params uploadop.CreateUploadParams) middlewa
 
 	userID, ok := authctx.GetUserID(params.HTTPRequest.Context())
 	if !ok {
-		h.logger.Panic("No User ID, this should never happen.")
```
Comment (ntwyman) [inline pkg/handlers/uploads.go]: This should h.logger.Error("Missing User ID in context");

File: pkg/handlers/uploads.go [hunk 6]
```diff
@@ -34,26 +34,37 @@ func (h CreateUploadHandler) Handle(params uploadop.CreateUploadParams) middlewa
...
+		return uploadop.NewCreateUploadBadRequest()
 	}
 
 	moveID, err := uuid.FromString(params.MoveID.String())
 	if err != nil {
-		h.logger.Panic("Invalid MoveID, this should never happen.")
```
Comment (ntwyman) [inline pkg/handlers/uploads.go]: This should 

`h.logger.Error("Badly formed UUID for moveId", zap.String(params.MoveID.String()), zap.Error(err))`

**Pull request merged**